src/compiler/scala/tools/nsc/ScriptRunner.scala

```diff
@@ -83,57 +87,69 @@ abstract class AbstractScriptRunner(settings: GenericRunnerSettings) extends Scr
    */
   private def withCompiledScript(scriptFile: String)(handler: String => Option[Throwable]): Option[Throwable] = {
 
-    /* Compiles the script file, and returns the directory with the compiled
-     * class files, if the compilation succeeded.
+    /* Compiles the script file, with the output set to either
+     * the user-specified location (jar or dir), or a temp dir.
+     * Returns the output location on success.
      */
-    def compile: Option[Directory] = {
-      val compiledPath = Directory.makeTemp("scalascript")
-
-      // delete the directory after the user code has finished
-      Runtime.getRuntime.addShutdownHook(new Thread(() => compiledPath.deleteRecursively()))
-
-      settings.outdir.value = compiledPath.path
-
-      if (doCompile(scriptFile)) Some(compiledPath) else None
+    def compile: Option[Path] = {
+      val outpath =
+        if (settings.outdir.isSetByUser)
+          Path(settings.outdir.value)
+        else
+          Directory.makeTemp("scalascript").tap { tmp =>
+            // delete the directory after the user code has finished
+            Runtime.getRuntime.addShutdownHook(new Thread(() => tmp.deleteRecursively()))
+            settings.outdir.value = tmp.path
+          }
+
+      if (doCompile(scriptFile)) Some(outpath) else None
     }
 
-    def hasClassToRun(d: Directory): Boolean = DirectoryClassPath(d.jfile).findClass(mainClass).isDefined
+    def hasClassToRun(location: Path): Boolean = {
+      val cp =
+        if (Jar.isJarOrZip(location)) {
+          import scala.reflect.io.AbstractFile
+          val fil = AbstractFile.getFile(location)
+          val reg = new CloseableRegistry
+          ZipAndJarClassPathFactory.create(fil, settings, reg)
+        }
+        else DirectoryClassPath(location.jfile)
```

Comment:
`ClassPathFactory.newClassPath`, and it'll even create the pesky `CloseableRegistry` for ya!